cmd/geth/snapshot.go

```diff
@@ -1020,3 +1034,155 @@ func checkAccount(ctx *cli.Context) error {
+	snaptree, err := snapshot.New(snapConfig, db, tdb, root, int(triesInMemory), false)
+	snap := snaptree.Snapshot(root)
+	if err != nil {
+		return err
+	}
+
+	// init key
+	keySet := make([]common.Hash, 0, 1000)
+	for i := 0; i < 1000; i++ {
+		key := common.BigToHash(big.NewInt(int64(i)))
+		keySet = append(keySet, crypto.Keccak256Hash(key.Bytes()))
+	}
+
+	cpuCount := ctx.Int64(utils.CPUCountFlag.Name)
+	p := NewRoutinePool(cpuCount*100, snap, keySet)
+
+	it := db.NewIterator(rawdb.SnapshotAccountPrefix, nil)
+	defer it.Release()
+
+	// Inspect key-value database first.
+	idx := uint64(0)
+	for it.Next() {
+		key := it.Key()
+		switch {
+		case bytes.HasPrefix(key, rawdb.SnapshotAccountPrefix) && len(key) == (len(rawdb.SnapshotAccountPrefix)+common.HashLength):
+			var account types.SlimAccount
+			err := rlp.DecodeBytes(it.Value(), &account)
+			if err != nil {
+				log.Error("failed to DecodeBytes", "addrHash", key[1:])
+				continue
+			}
+			if bytes.Equal(account.CodeHash, types.EmptyCodeHash.Bytes()) {
+				continue
+			}
+			fmt.Printf("add task: %d\n", idx)
+			idx++
+			p.AddTask(common.BytesToHash(key[1:]))
+		}
+	}
+
+	p.Wait()
+	fmt.Println("all done")
+	fmt.Println("var count", p.varCount.Load())
+	fmt.Println("arr count", p.arrCount.Load())
+	fmt.Println("arr item count", p.arrItemCount.Load())
+
+	f, err := os.Create("result.txt")
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	_, err = f.WriteString(fmt.Sprintf("var count: %d\n", p.varCount.Load()))
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	_, err = f.WriteString(fmt.Sprintf("arr count: %d\n", p.arrCount.Load()))
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	_, err = f.WriteString(fmt.Sprintf("arr item count: %d\n", p.arrItemCount.Load()))
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	return nil
+}
+
+func traversalContract(snap snapshot.Snapshot, contractAddress common.Hash, keySet []common.Hash) (int, int, int) {
+	emptyKeyCount := 0
+	varCount := 0
+	arrayCount := 0
+	arrayVarCount := 0
+
+	for _, storageHash := range keySet {
+		enc, encErr := snap.Storage(contractAddress, storageHash)
+		if encErr != nil {
+			log.Info(fmt.Sprintf("key: %s, storage error: %v", storageHash.String(), encErr))
+
+			continue
+		}
+		if len(enc) > 0 {
+			emptyKeyCount = 0
+			if isArr, arrLen := isArray(snap, contractAddress, storageHash); isArr {
+				arrayVarCount += arrLen
+				arrayCount++
+			} else {
+				varCount++
+			}
+
+		} else {
+			emptyKeyCount++
+		}
+		if emptyKeyCount == 20 {
+			break
+		}
+	}
+	//log.Info("array count:", arrayCount, "array item count", arrayVarCount, "var count", varCount)
+
+	return arrayCount, arrayVarCount, varCount
+}
+
+func isArray(snap snapshot.Snapshot, contractAddress common.Hash, slotIdx common.Hash) (result bool, slotLen int) {
+	emptyValue := 0
+	hasValue := false
+	errTime := 0
+	for i := 0; ; i++ {
+		if errTime > 100 {
+			break
+		}
+		idx := common.BigToHash(big.NewInt(0).Add(slotIdx.Big(), big.NewInt(int64(i))))
```

Comment:
Which source did you reference for this code?

Reply:
No source

Reply:
No problem with the test